src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Services/ExtensionAssemblyManager.cs

```diff
@@ -0,0 +1,151 @@
+            if (fileNameNoExt == null)
+            {
+                logger.LogError("Failed to get file name without extension from {assemblyFilePath}", assemblyFilePath);
+                continue;
+            }
+
+            if (directoryLoadContexts.TryGetValue(directory, out var _))
+            {
+                logger.LogTrace("Reusing existing load context for {assemblyFilePath}", assemblyFilePath);
+                continue;
+            }
+
+            // Create the extension assembly load context for the extension.
+            logger.LogTrace("Loading extension {assemblyFilePath}", assemblyFilePath);
+            var loadContext = new ExtensionAssemblyLoadContext(fileNameNoExt, directory, loggerFactory);
+            directoryLoadContexts.Add(directory, loadContext);
+        }
+
+        return new ExtensionAssemblyManager(directoryLoadContexts.ToImmutableDictionary());
+    }
+
+    private static void LoadDevKitAssemblies(string devKitDependencyPath, ILogger logger)
+    {
+        var directoryName = Path.GetDirectoryName(devKitDependencyPath);
+        Contract.ThrowIfNull(directoryName);
+        logger.LogTrace("Loading DevKit assemblies from {directory}", directoryName);
+
+        var directory = new DirectoryInfo(directoryName);
+        foreach (var file in directory.GetFiles("*.dll"))
+        {
+            logger.LogTrace("Loading {assemblyName}", file.Name);
+            // DevKit assemblies are loaded into the default load context. This allows extensions
+            // to share the host's instance of these assemblies as long as they do not ship their own copy.
+            AssemblyLoadContext.Default.LoadFromAssemblyPath(file.FullName);
+        }
+    }
+
+    /// <summary>
+    /// Loads an assembly from an assembly file path into the extension load context for the assembly's directory.
+    /// If the directory containing the assembly file path is not an extension directory, this will return null.
+    /// </summary>
+    public Assembly? TryLoadAssemblyInExtensionContext(string assemblyFilePath)
+    {
+        var directory = Path.GetDirectoryName(assemblyFilePath);
+        if (directory == null)
+        {
+            return null;
+        }
+
+        if (_directoryLoadContexts.TryGetValue(directory, out var loadContext))
+        {
+            return loadContext.LoadFromAssemblyPath(assemblyFilePath);
+        }
+
+        return null;
+    }
+
+    /// <summary>
+    /// Loads an assembly from an assembly name in the extension load context.
+    /// This will attempt to load the assembly from each extension load context and use the first one that succeeds.
+    /// </summary>
+    public Assembly? TryLoadAssemblyInExtensionContext(AssemblyName assemblyName)
+    {
+        // We don't know exactly which extension the assembly came from, so we'll try each extension load context.
+        foreach (var loadContext in _directoryLoadContexts.Values)
```

Comment:
cc @jasonmalinowski  - I wasn't sure if there was a better way to do this.  Basically the MEF custom assembly loader calls into us with an `AssemblyName`.  I'm not sure how to get the actual path it is trying to load to find the exact extension ALC to use.  There is a codebase path on `AssemblyName`, but its obsolete.

Reply:
@jasonmalinowski hmm I could do the assembly name map assuming that the assembly name from extension a and b are different.  

> It's a bit unfortunate though that VS MEF can't give us a full path in this case?
The code base does have the full path in this case, but its marked obsolete.  I could suppress that warning and check it anyway?

Reply:
> The code base does have the full path in this case, but its marked obsolete. I could suppress that warning and check it anyway?

Oh I see: I was imagining you'd use the codebase given to the VS MEF IAssemblyLoader call that does have one, if it's available. But practically the only way we can ever make this work is if VS MEF is reliably able to tells the actual underlying path. If it's taking a model that you can't have more than one extension w/ the same assembly name then that might just be a limitation we have. And I can imagine that a least on VS for Windows, that's a reasonable limitation given the underlying .NET Framework limitations.

Reply:
MEF in general (VS-MEF or otherwise) makes a very huge assumption that type full names (without their assembly names to qualify them) are unique in the MEF catalog. By default (and this comprises 99% of the cases) an import or export will use the `Type.FullName` as the MEF contract name, which is how imports and exports are matched. 
If the same assembly is included in the catalog twice (whether or not it's the same version of the assembly) and that assembly includes MEF parts, it will most likely produce _many_ cases of export duplication, which would bring down everyone who expected just one export (i.e. `[Import]`) and those who use `[ImportMany]` would get an export from each assembly. 
You should therefore be sure to only include _one_ assembly of a given name in your MEF catalog. You might choose the newest version if there are multiple to choose from.
If you include _one_ in the MEF catalog but you allow other copies to also load (in other ALCs), that shouldn't confuse _MEF_, but it could still produce some confusing errors around type non-equivalence, and some of those errors _could_ show up with MEF on the callstack.

Reply:
> You should therefore be sure to only include one assembly of a given name in your MEF catalog. You might choose the newest version if there are multiple to choose from.
If you include one in the MEF catalog but you allow other copies to also load (in other ALCs), that shouldn't confuse MEF, but it could still produce some confusing errors around type non-equivalence, and some of those errors could show up with MEF on the callstack.

Yeah that was my rough plan - we'll only include an assembly once into the catalog, but we'll allow an extension assembly to have other copies in its own ALC that do not contribute to the catalog.

Reply:
Did you still want to do a map here so we're not having to do an O(n) loop?